<?xml version="1.0" encoding="utf-8" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="Docutils 0.8.1: http://docutils.sourceforge.net/" />
<title>color-KEGG-pathways</title>
<style type="text/css">

/*
:Author: David Goodger (goodger@python.org)
:Id: $Id: html4css1.css 7056 2011-06-17 10:50:48Z milde $
:Copyright: This stylesheet has been placed in the public domain.

Default cascading style sheet for the HTML output of Docutils.

See http://docutils.sf.net/docs/howto/html-stylesheets.html for how to
customize this style sheet.
*/

/* used to remove borders from tables and images */
.borderless, table.borderless td, table.borderless th {
  border: 0 }

table.borderless td, table.borderless th {
  /* Override padding for "table.docutils td" with "! important".
     The right padding separates the table cells. */
  padding: 0 0.5em 0 0 ! important }

.first {
  /* Override more specific margin styles with "! important". */
  margin-top: 0 ! important }

.last, .with-subtitle {
  margin-bottom: 0 ! important }

.hidden {
  display: none }

a.toc-backref {
  text-decoration: none ;
  color: black }

blockquote.epigraph {
  margin: 2em 5em ; }

dl.docutils dd {
  margin-bottom: 0.5em }

object[type="image/svg+xml"], object[type="application/x-shockwave-flash"] {
  overflow: hidden;
}

/* Uncomment (and remove this text!) to get bold-faced definition list terms
dl.docutils dt {
  font-weight: bold }
*/

div.abstract {
  margin: 2em 5em }

div.abstract p.topic-title {
  font-weight: bold ;
  text-align: center }

div.admonition, div.attention, div.caution, div.danger, div.error,
div.hint, div.important, div.note, div.tip, div.warning {
  margin: 2em ;
  border: medium outset ;
  padding: 1em }

div.admonition p.admonition-title, div.hint p.admonition-title,
div.important p.admonition-title, div.note p.admonition-title,
div.tip p.admonition-title {
  font-weight: bold ;
  font-family: sans-serif }

div.attention p.admonition-title, div.caution p.admonition-title,
div.danger p.admonition-title, div.error p.admonition-title,
div.warning p.admonition-title {
  color: red ;
  font-weight: bold ;
  font-family: sans-serif }

/* Uncomment (and remove this text!) to get reduced vertical space in
   compound paragraphs.
div.compound .compound-first, div.compound .compound-middle {
  margin-bottom: 0.5em }

div.compound .compound-last, div.compound .compound-middle {
  margin-top: 0.5em }
*/

div.dedication {
  margin: 2em 5em ;
  text-align: center ;
  font-style: italic }

div.dedication p.topic-title {
  font-weight: bold ;
  font-style: normal }

div.figure {
  margin-left: 2em ;
  margin-right: 2em }

div.footer, div.header {
  clear: both;
  font-size: smaller }

div.line-block {
  display: block ;
  margin-top: 1em ;
  margin-bottom: 1em }

div.line-block div.line-block {
  margin-top: 0 ;
  margin-bottom: 0 ;
  margin-left: 1.5em }

div.sidebar {
  margin: 0 0 0.5em 1em ;
  border: medium outset ;
  padding: 1em ;
  background-color: #ffffee ;
  width: 40% ;
  float: right ;
  clear: right }

div.sidebar p.rubric {
  font-family: sans-serif ;
  font-size: medium }

div.system-messages {
  margin: 5em }

div.system-messages h1 {
  color: red }

div.system-message {
  border: medium outset ;
  padding: 1em }

div.system-message p.system-message-title {
  color: red ;
  font-weight: bold }

div.topic {
  margin: 2em }

h1.section-subtitle, h2.section-subtitle, h3.section-subtitle,
h4.section-subtitle, h5.section-subtitle, h6.section-subtitle {
  margin-top: 0.4em }

h1.title {
  text-align: center }

h2.subtitle {
  text-align: center }

hr.docutils {
  width: 75% }

img.align-left, .figure.align-left, object.align-left {
  clear: left ;
  float: left ;
  margin-right: 1em }

img.align-right, .figure.align-right, object.align-right {
  clear: right ;
  float: right ;
  margin-left: 1em }

img.align-center, .figure.align-center, object.align-center {
  display: block;
  margin-left: auto;
  margin-right: auto;
}

.align-left {
  text-align: left }

.align-center {
  clear: both ;
  text-align: center }

.align-right {
  text-align: right }

/* reset inner alignment in figures */
div.align-right {
  text-align: inherit }

/* div.align-center * { */
/*   text-align: left } */

ol.simple, ul.simple {
  margin-bottom: 1em }

ol.arabic {
  list-style: decimal }

ol.loweralpha {
  list-style: lower-alpha }

ol.upperalpha {
  list-style: upper-alpha }

ol.lowerroman {
  list-style: lower-roman }

ol.upperroman {
  list-style: upper-roman }

p.attribution {
  text-align: right ;
  margin-left: 50% }

p.caption {
  font-style: italic }

p.credits {
  font-style: italic ;
  font-size: smaller }

p.label {
  white-space: nowrap }

p.rubric {
  font-weight: bold ;
  font-size: larger ;
  color: maroon ;
  text-align: center }

p.sidebar-title {
  font-family: sans-serif ;
  font-weight: bold ;
  font-size: larger }

p.sidebar-subtitle {
  font-family: sans-serif ;
  font-weight: bold }

p.topic-title {
  font-weight: bold }

pre.address {
  margin-bottom: 0 ;
  margin-top: 0 ;
  font: inherit }

pre.literal-block, pre.doctest-block, pre.math {
  margin-left: 2em ;
  margin-right: 2em }

span.classifier {
  font-family: sans-serif ;
  font-style: oblique }

span.classifier-delimiter {
  font-family: sans-serif ;
  font-weight: bold }

span.interpreted {
  font-family: sans-serif }

span.option {
  white-space: nowrap }

span.pre {
  white-space: pre }

span.problematic {
  color: red }

span.section-subtitle {
  /* font-size relative to parent (h1..h6 element) */
  font-size: 80% }

table.citation {
  border-left: solid 1px gray;
  margin-left: 1px }

table.docinfo {
  margin: 2em 4em }

table.docutils {
  margin-top: 0.5em ;
  margin-bottom: 0.5em }

table.footnote {
  border-left: solid 1px black;
  margin-left: 1px }

table.docutils td, table.docutils th,
table.docinfo td, table.docinfo th {
  padding-left: 0.5em ;
  padding-right: 0.5em ;
  vertical-align: top }

table.docutils th.field-name, table.docinfo th.docinfo-name {
  font-weight: bold ;
  text-align: left ;
  white-space: nowrap ;
  padding-left: 0 }

h1 tt.docutils, h2 tt.docutils, h3 tt.docutils,
h4 tt.docutils, h5 tt.docutils, h6 tt.docutils {
  font-size: 100% }

ul.auto-toc {
  list-style-type: none }

</style>
</head>
<body>
<div class="document" id="color-kegg-pathways">
<h1 class="title">color-KEGG-pathways</h1>

<p><strong>color-KEGG-pathways</strong> is a command-line utility which interfaces with the <a class="reference external" href="http://www.genome.jp/kegg/">KEGG</a> <a class="reference external" href="http://www.genome.jp/kegg/soap/doc/keggapi_manual.html">API</a> to produce color-coded figures of pathways. Any enzyme, metabolite or sub-pathway normally visible in a KEGG pathway can be painted with arbitrary color to represent quantitative or qualitative information such as expression level, presence or absence, etc. Here is an example:</p>
<div class="figure align-center">
<img alt="Example of color-coded pathway" src="example.png" />
<p class="caption">Example of color-coded pathway.</p>
<div class="legend">
Enzymes in the glycolysis pathway have been colored according to experimentally measured gene expression levels. The color of each enzyme is calculated using a 3-point gradient of yellow (for the most up-regulated gene) to black (no expression) to blue (most down-regulated gene). White color is used to represent enzymes that are not encoded by any gene in this species.</div>
</div>
<p>Information about which objects to color and how to color them must be provided to <strong>color-KEGG-pathways</strong> as a <a class="reference external" href="http://en.wikipedia.org/wiki/Comma-separated_values">.csv</a>-formatted input file with the following syntax:</p>
<ul class="simple">
<li>the first column must be the KEGG identifier for this pathway (e.g., <a class="reference external" href="http://www.genome.jp/kegg-bin/show_pathway?sce00010">sce00010</a> for the Saccharomyces cerevisiae glycolysis pathway)</li>
<li>the second column must be the KEGG identifier of an object in this pathway (e.g., <a class="reference external" href="http://www.genome.jp/dbget-bin/www_bget?sce:YBR019C">YBR019C</a> for enzymes 5.1.3.2 and 5.1.3.3)</li>
<li>the third column must either be a HTML-compatible <a class="reference external" href="http://en.wikipedia.org/wiki/Web_colors#Hex_triplet">hex triplet</a> color (e.g., <em>#FFFFFF</em> for white or <em>#0000FF</em> for blue) or a measure. In this later case several options are available for <strong>color-KEGG-pathways</strong> to calculate the color based on overall maximum and minimum values</li>
</ul>
<p>For example, the .csv file used to obtain the figure above was:</p>
<pre class="literal-block">
afm00010,afm:AFUA_5G01030,1.82653195
afm00010,afm:AFUA_3G05740,1.7700902
afm00010,afm:AFUA_3G09290,1.1786362
afm00010,afm:AFUA_3G11070,1.0063019
afm00010,afm:AFUA_5G14810,0.5802401
afm00010,afm:AFUA_5G06240,0.49131745
afm00010,afm:AFUA_2G04870,0.40439728
afm00010,afm:AFUA_8G02560,0.2803983
afm00010,afm:AFUA_4G08600,0.209255325
afm00010,afm:AFUA_3G11900,0.207820625
</pre>
<p>(note: only the first ten lines are shown here)</p>
<p>More than one pathway can be referred to in the same input file; <strong>color-KEGG-pathways</strong> will process each pathway and generate individual pictures.</p>
<div class="section" id="latest-version">
<h1>Latest version</h1>
<p>The current version of <strong>color-KEGG-pathways</strong> is VERSION.</p>
</div>
<div class="section" id="installation">
<h1>Installation</h1>
<p>To install <strong>color-KEGG-pathways</strong> you must go through the following steps:</p>
<ul class="simple">
<li>ensure a Python interpreter (version 2.6 or 2.7; 3.0 and later are not supported yet) is installed in your system. Please refer to this <a class="reference external" href="http://www.python.org/getit/">page</a> if needed.</li>
<li>ensure the <a class="reference external" href="http://fedorahosted.org/suds/">SUDS</a>, <a class="reference external" href="http://pyx.sourceforge.net/">PyX</a> and <a class="reference external" href="http://www.pythonware.com/products/pil/">PIL</a> libraries are installed. A quick way to install those is to first install <em>easy_install</em> (see the <a class="reference external" href="http://pypi.python.org/pypi/setuptools">documentation</a>) then type <em>easy_install suds</em>, <em>easy_install pyx</em> then <em>easy_install pil</em> on the command line.</li>
<li>go to <a class="reference external" href="http://github.com/ajmazurie/Utilities/zipball/master">http://github.com/ajmazurie/Utilities/zipball/master</a>. A file named <em>ajmazurie-Utilities-xxxx.zip</em> will download (xxxx will be replaced by a commit signature code).</li>
<li>unzip this file.</li>
<li>ensure the file <em>color-KEGG-pathways/color-KEGG-pathways</em> is in your <a class="reference external" href="http://kb.iu.edu/data/acar.html">PATH</a>, either by moving it to a directory already in your PATH or adding this directory to the PATH.</li>
<li><strong>color-KEGG-pathways</strong> should now be accessible from the command line; you can test it by typing <em>color-KEGG-pathways --version</em></li>
</ul>
</div>
<div class="section" id="usage">
<h1>Usage</h1>
<p>A list of options can be obtained by using the <em>--help</em> option:</p>
<pre class="literal-block">
%&gt; color-KEGG-pathways --help
Usage: color-KEGG-pathways [options]
</pre>
</div>
<div class="section" id="examples">
<h1>Examples</h1>
<div class="section" id="example-1">
<h2>Example 1</h2>
<p>Coloring the enzymes of the <em>Saccharomyces cerevisiae</em> citrate/TCA cycle (pathway identifier <em>sce00020</em>) with a rainbow pattern, and the metabolites in red.</p>
<p>Content of <a class="reference external" href="example_1.csv">example_1.csv</a>:</p>
<pre class="literal-block">
sce00020,sce:YCR005C,ff0000
sce00020,sce:YNR001C,ff1f00
sce00020,sce:YPR001W,ff3f00
sce00020,sce:YJL200C,ff5f00
sce00020,sce:YLR304C,ff7f00
sce00020,sce:YDL066W,ff9f00
sce00020,sce:YLR174W,ffbf00
sce00020,sce:YNL009W,ffdf00
sce00020,sce:YNL037C,ffff00
sce00020,sce:YOR136W,dfff00
sce00020,sce:YIL125W,bfff00
sce00020,sce:YDR148C,9fff00
sce00020,sce:YFL018C,7fff00
sce00020,sce:YPL017C,5fff00
sce00020,sce:YOR142W,3fff00
sce00020,sce:YGR244C,1fff00
sce00020,sce:YJL045W,00ff00
sce00020,sce:YKL148C,00ff1f
sce00020,sce:YLL041C,00ff3f
sce00020,sce:YKL141W,00ff5f
sce00020,sce:YMR118C,00ff7f
sce00020,sce:YDR178W,00ff9f
sce00020,sce:YLR164W,00ffbf
sce00020,sce:YPL262W,00ffdf
sce00020,sce:YDL078C,00ffff
sce00020,sce:YKL085W,00dfff
sce00020,sce:YOL126C,00bfff
sce00020,sce:YBR218C,009fff
sce00020,sce:YGL062W,007fff
sce00020,sce:YKR097W,005fff
sce00020,sce:YER178W,003fff
sce00020,sce:YBR221C,001fff
sce00020,sce:YNL071W,0000ff
sce00020,cpd:C00022,ff0000
sce00020,cpd:C00024,ff0000
sce00020,cpd:C00026,ff0000
sce00020,cpd:C00036,ff0000
sce00020,cpd:C00042,ff0000
sce00020,cpd:C00068,ff0000
sce00020,cpd:C00074,ff0000
sce00020,cpd:C00091,ff0000
sce00020,cpd:C00122,ff0000
sce00020,cpd:C00149,ff0000
sce00020,cpd:C00158,ff0000
sce00020,cpd:C00311,ff0000
sce00020,cpd:C00417,ff0000
sce00020,cpd:C05125,ff0000
sce00020,cpd:C05379,ff0000
sce00020,cpd:C05381,ff0000
sce00020,cpd:C15972,ff0000
sce00020,cpd:C15973,ff0000
sce00020,cpd:C16254,ff0000
</pre>
<p>Command-line syntax:</p>
<pre class="literal-block">
color-KEGG-pathways -i example_1.csv
</pre>
<p>Resulting picture:</p>
<img alt="example_1.png" class="align-center" src="example_1.png" />
</div>
<div class="section" id="example-2">
<h2>Example 2</h2>
<p>You will notice that the figure from Example 1 has gray boxes on some elements: namely, some enzymes and all of the connected pathways. This is because by default <strong>color-KEGG-pathways</strong> will assign a color (which you can change using the <em>--missing-color</em> option) to missing objects; i.e., objects that are part of the pathway but are not listed in the input .csv file.</p>
<p>You can change this behavior by using the <em>--dont-show-missing</em> and <em>--only-show-missing</em> options. The first one will not show missing objects of some type (compound, genes, ortholog or map) while the second will only show missing objects of a given type.</p>
<p>Hence, to create a new version of the figure above with only missing genes being colored in gray you can type:</p>
<pre class="literal-block">
color-KEGG-pathways -i example_1.csv --only-show-missing genes
</pre>
<p>Resulting picture:</p>
<img alt="example_2.png" class="align-center" src="example_2.png" />
<p>In this case no missing gene can be found. Note that white rectangles are a way for KEGG to show that the genes corresponding to a given enzyme are not found in this particular species.</p>
</div>
<div class="section" id="example-3">
<h2>Example 3</h2>
<p>Let now use another input file, in which genes are not provided with colors but with some measurements, such as expression ratios. By using the <em>--two-color-scale</em> option <strong>color-KEGG-pathways</strong> will set the color of these genes as an interpolation between a <em>start</em> and an <em>end</em> color (which can be modified using the <em>--start-color</em> and <em>--end-color</em> options, respectively). <em>--two-color-scale</em> must be provided with the measure associated to the start color and the stop color; the <em>min</em> and <em>max</em> placeholders can be used to represent the minimum and maximum value in the input file, respectively.</p>
<p>Content of <a class="reference external" href="example_3.csv">example_3.csv</a>:</p>
<pre class="literal-block">
sce00020,sce:YCR005C,0.126282121306
sce00020,sce:YNR001C,0.732643560422
sce00020,sce:YPR001W,0.451466310516
sce00020,sce:YJL200C,0.63541088878
sce00020,sce:YLR304C,0.29984681721
sce00020,sce:YDL066W,0.686157749236
sce00020,sce:YLR174W,0.184382431261
sce00020,sce:YNL009W,0.503979008833
sce00020,sce:YNL037C,0.313627903323
sce00020,sce:YOR136W,0.738991695825
sce00020,sce:YIL125W,0.995497433424
sce00020,sce:YDR148C,0.976824618511
sce00020,sce:YFL018C,0.221671234839
sce00020,sce:YPL017C,0.496774542881
sce00020,sce:YOR142W,0.753411521095
sce00020,sce:YGR244C,0.871886510744
sce00020,sce:YJL045W,0.63206417879
sce00020,sce:YKL148C,0.635856811757
sce00020,sce:YLL041C,0.866004577703
sce00020,sce:YKL141W,0.559150167364
sce00020,sce:YMR118C,0.165359091514
sce00020,sce:YDR178W,0.5630501412
sce00020,sce:YLR164W,0.110842747088
sce00020,sce:YPL262W,0.47728178047
sce00020,sce:YDL078C,0.339389017326
sce00020,sce:YKL085W,0.643389560396
sce00020,sce:YOL126C,0.810766941838
sce00020,sce:YBR218C,0.128383581285
sce00020,sce:YGL062W,0.305526154636
sce00020,sce:YKR097W,0.96793319272
sce00020,sce:YER178W,0.354643109199
sce00020,sce:YBR221C,0.899168181625
sce00020,sce:YNL071W,0.790556669728
</pre>
<p>Command-line syntax:</p>
<pre class="literal-block">
color-KEGG-pathways -i example_3.csv --two-color-scale min max --only-show-missing genes
</pre>
<p>Resulting picture:</p>
<img alt="example_3.png" class="align-center" src="example_3.png" />
<p>In this example the gene with the minimum measure received the color set by <em>--start-color</em>, while the gene with the maximum measure received the color set by <em>--end-color</em>. All the genes in between received a color interpolated between these two colors.</p>
<p>Note that the picture now has a legend on the left, showing the gradient between the start and stop color and the associated measures. You can remove this legend by using the <em>--no-legend</em> option.</p>
</div>
<div class="section" id="example-4">
<h2>Example 4</h2>
<p>In this last example we will show how to use a 3-color scale. This is particularly useful to represent expression ratio, which are either positive or negative. In this case using two color gradients make sense: one gradient for all values between a midpoint (typically zero) and the maximum ratio, and another gradient for all values between the midpoint and the minimum ratio.</p>
<p>This can be done using the <em>--three-color-scale</em> option. Similarly to <em>--two-color-scale</em> it must be provided with the values of the start, midpoint and stop colors.</p>
<p>In addition to the <em>--start-color</em> and <em>--end-color</em> options seen above, you can use the <em>--midpoint-color</em> to set the color of the midpoint.</p>
<p>Content of <a class="reference external" href="example_4.csv">example_4.csv</a>:</p>
<pre class="literal-block">
Dummy
</pre>
<p>Command-line syntax:</p>
<pre class="literal-block">
color-KEGG-pathways -i example_4.csv --three-color-scale min 0 max --only-show-missing genes
</pre>
<p>Resulting picture:</p>
<img alt="example_4.png" class="align-center" src="example_4.png" />
</div>
</div>
<div class="section" id="note-about-colors">
<h1>Note about colors</h1>
<p>The use of green for down-regulated and green for up-regulated genes and proteins is a <em>de facto</em> standard in the literature. Those colors, however, are difficult to distinguish for most color-blind people.</p>
<p>As discussed at <a class="reference external" href="http://jfly.iam.u-tokyo.ac.jp/color/">http://jfly.iam.u-tokyo.ac.jp/color/</a> the following replacements are better options:</p>
<ul class="simple">
<li>instead of green (0 255 0) and red (255 0 0) you should prefer bluish green (0 158 115) and orange (230 159 0)</li>
<li>even better, you should prefer sky blue (86 180 233) and yellow (240 228 66); those have a higher contrast than the previous choice</li>
</ul>
</div>
</div>
</body>
</html>

